### ISSUE TITLE
[Facebook Live]I can't receive the stream when I use 4g (Cellular data) instead of Wifi

### ISSUE THREAD

**Issue by geitoodevs** [shogo4405/lf.swift#83]
Only when I turn of the Wifi connection I get the stream published and emitting properly.

Do you know how can I enable this? or Why is this happening?
I checked the settings of the app and Data Mobile is turned on.

Thanks, regards!
Miguel.

**Comment by shogo4405** [comment]
Can I see your application log?
Please set XCGLogger

    import XCGLogger
    
    XCGLogger.defaultInstance().setup(
            .Debug,
            showThreadName: true, showLogLevel: true, showFileNames: true, showLineNumbers: true, writeToFile: nil, fileLogLevel: nil)

**Comment by shogo4405** [comment]
By the way. talking about RTMPStream?

**Comment by geitoodevs** [comment]
Yes, this is the log when I connect the device with WIFI and after this call:
        rtmpConnection.connect(url!)
        rtmpStream.publish(stream)

**2016-08-03 07:07:13.408 [Debug] [main] [VideoIOComponent.swift:75] VideoIOComponent > cicontext use hardware renderer
2016-08-03 07:07:13.914 [Info] [com.github.shogo4405.lf.AudioIOComponent.lock] [AACEncoder.swift:78] inSourceFormat > nil
2016-08-03 07:07:14.009 [Info] [com.github.shogo4405.lf.Stream.lock] [VideoIOComponent.swift:94] fps > (30.0, __C.CMTime(value: 100, timescale: 3000, flags: __C.CMTimeFlags(rawValue: 1), epoch: 0))
warning: could not load any Objective-C class information from the dyld shared cache. This will significantly reduce the quality of type information available.
2016-08-03 07:08:00.129 [Info] [com.apple.avfoundation.audiodataoutput.bufferqueue] [AACEncoder.swift:78] inSourceFormat > Optional(__C.AudioStreamBasicDescription(mSampleRate: 44100.0, mFormatID: 1819304813, mFormatFlags: 12, mBytesPerPacket: 2, mFramesPerPacket: 1, mBytesPerFrame: 2, mChannelsPerFrame: 1, mBitsPerChannel: 16, mReserved: 0))**

At this moment when I check the video streaming, it's working properly. But when I turn off the Wifi and the device is using 4G the log is:

**2016-08-03 07:09:26.896 [Debug] [main] [VideoIOComponent.swift:75] VideoIOComponent > cicontext use hardware renderer
2016-08-03 07:09:27.208 [Info] [com.github.shogo4405.lf.AudioIOComponent.lock] [AACEncoder.swift:78] inSourceFormat > nil
2016-08-03 07:09:27.279 [Info] [com.github.shogo4405.lf.Stream.lock] [VideoIOComponent.swift:94] fps > (30.0, __C.CMTime(value: 100, timescale: 3000, flags: __C.CMTimeFlags(rawValue: 1), epoch: 0))**

And nothing is shown in the preview url of the streaming video.

My configuration is:
         rtmpStream = RTMPStream(rtmpConnection: rtmpConnection)
    
        let sampleRate:Double = 44_100
        do {
            try AVAudioSession.sharedInstance().setPreferredSampleRate(sampleRate)
            try AVAudioSession.sharedInstance().setCategory(AVAudioSessionCategoryPlayAndRecord)
            try AVAudioSession.sharedInstance().setMode(AVAudioSessionModeVideoChat)
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
        }
        rtmpStream.audioSettings = [
            "sampleRate": sampleRate,
        ]
	   rtmpStream.attachAudio(AVCaptureDevice.defaultDeviceWithMediaType(AVMediaTypeAudio), false)
        
        rtmpStream.videoSettings = [
            "width": 320, // video output width
            "height": 320, // video output height
            "bitrate": 256 * 1024, // video output bitrate
            "profileLevel": kVTProfileLevel_H264_Baseline_3_1
        ]
        rtmpStream.view.frame = videoContainerView.frame
        rtmpStream.view.videoGravity = AVLayerVideoGravityResizeAspectFill
        let device = AVMixer.deviceWithPosition(.Front)
        rtmpStream.attachCamera(device)

I don't know why is working right with Wifi but not for 4g (I have cellular data internet connection)

**Comment by shogo4405** [comment]
It seems settings OK. I think network problem... RTMP port is 1935.
Can u access to your server from external network (4G) How about Youtube Live? 
```swift
XCGLogger.defaultInstance().setup(
        .Verbose, // change .Debug to .Verbose
        showThreadName: true, showLogLevel: true, showFileNames: true, showLineNumbers: true, writeToFile: nil, fileLogLevel: nil)
       

rtmpConnection.connect("rtmp://a.rtmp.youtube.com/live2")
rtmpStream.publish("live"); // dummy
```
This log means. 
* RTMP Handshake is done.
* RTMP Connection is up.
* RTMPStream auth is failed.
```swift
2016-08-03 18:10:24.688 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPSocket.swift:65] doOutput(chunk:) > RTMPChunk{size:0,type:Zero,streamId:3,message:Optional(RTMPCommandMessage{type:AMF0Command,length:276,streamId:0,timestamp:0,payload:[2, 0, 7, 99, 111, 110, 110, 101, 99, 116, 0, 63, 240, 0, 0, 0, 0, 0, 0, 3, 0, 5, 116, 99, 85, 114, 108, 2, 0, 31, 114, 116, 109, 112, 58, 47, 47, 97, 46, 114, 116, 109, 112, 46, 121, 111, 117, 116, 117, 98, 101, 46, 99, 111, 109, 47, 108, 105, 118, 101, 50, 0, 8, 102, 108, 97, 115, 104, 86, 101, 114, 2, 0, 31, 70, 77, 76, 69, 47, 51, 46, 48, 32, 40, 99, 111, 109, 112, 97, 116, 105, 98, 108, 101, 59, 32, 70, 77, 83, 99, 47, 49, 46, 48, 41, 0, 6, 115, 119, 102, 85, 114, 108, 5, 0, 3, 97, 112, 112, 2, 0, 5, 108, 105, 118, 101, 50, 0, 4, 102, 112, 97, 100, 1, 0, 0, 11, 97, 117, 100, 105, 111, 67, 111, 100, 101, 99, 115, 0, 64, 144, 0, 0, 0, 0, 0, 0, 0, 11, 118, 105, 100, 101, 111, 67, 111, 100, 101, 99, 115, 0, 64, 96, 0, 0, 0, 0, 0, 0, 0, 13, 118, 105, 100, 101, 111, 70, 117, 110, 99, 116, 105, 111, 110, 0, 63, 240, 0, 0, 0, 0, 0, 0, 0, 12, 99, 97, 112, 97, 98, 105, 108, 105, 116, 105, 101, 115, 0, 64, 109, 224, 0, 0, 0, 0, 0, 0, 7, 112, 97, 103, 101, 85, 114, 108, 5, 0, 14, 111, 98, 106, 101, 99, 116, 69, 110, 99, 111, 100, 105, 110, 103, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 9],objectEncoding:0,commandName:connect,transactionId:1,commandObject:Optional(["tcUrl": Optional("rtmp://a.rtmp.youtube.com/live2"), "flashVer": Optional("FMLE/3.0 (compatible; FMSc/1.0)"), "swfUrl": nil, "app": Optional("live2"), "fpad": Optional(false), "audioCodecs": Optional(1024), "videoCodecs": Optional(128), "videoFunction": Optional(1), "capabilities": Optional(239), "pageUrl": nil, "objectEncoding": Optional(0)]),arguments:[],serializer:AMF0Serializer{bytes:[],position:0,reference:lf.AMFReference}}),fragmented:false,_bytes:[3, 0, 0, 0, 0, 1, 8, 20, 0, 0, 0, 0]}[;
[fg169,169,169;[bg;2016-08-03 18:10:24.772 [Debug] [com.github.shogo4405.lf.NetSocket.network] [NetSocket.swift:139] stream(_:handleEvent:) > eventCode: NSStreamEvent(rawValue: 4)[;
[fg169,169,169;[bg;2016-08-03 18:10:32.709 [Debug] [com.github.shogo4405.lf.NetSocket.network] [NetSocket.swift:139] stream(_:handleEvent:) > eventCode: NSStreamEvent(rawValue: 2)[;
[fg211,211,211;[bg;2016-08-03 18:10:32.714 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPConnection.swift:405] listen(_:bytes:) > RTMPChunk{size:128,type:Zero,streamId:2,message:Optional(RTMPWindowAcknowledgementSizeMessage{type:WindowAck,length:4,streamId:0,timestamp:0,payload:[0, 38, 37, 160],size:2500000}),fragmented:false,_bytes:[2, 0, 0, 0, 0, 0, 4, 5, 0, 0, 0, 0]}[;
[fg211,211,211;[bg;2016-08-03 18:10:32.716 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPSocket.swift:65] doOutput(chunk:) > RTMPChunk{size:0,type:Zero,streamId:2,message:Optional(RTMPWindowAcknowledgementSizeMessage{type:WindowAck,length:16,streamId:0,timestamp:0,payload:[0, 38, 37, 160],size:2500000}),fragmented:false,_bytes:[2, 0, 0, 0, 0, 0, 4, 5, 0, 0, 0, 0]}[;
[fg211,211,211;[bg;2016-08-03 18:10:32.718 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPConnection.swift:405] listen(_:bytes:) > RTMPChunk{size:128,type:One,streamId:2,message:Optional(RTMPSetPeerBandwidthMessage{type:Bandwidth,length:5,streamId:0,timestamp:0,payload:[0, 152, 150, 128, 2],size:0,limit:Hard}),fragmented:false,_bytes:[66, 0, 0, 0, 0, 0, 5, 6]}[;
[fg211,211,211;[bg;2016-08-03 18:10:33.395 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPConnection.swift:405] listen(_:bytes:) > RTMPChunk{size:128,type:One,streamId:2,message:Optional(RTMPSetChunkSizeMessage{type:ChunkSize,length:4,streamId:0,timestamp:0,payload:[0, 0, 1, 0],size:256}),fragmented:false,_bytes:[66, 0, 0, 0, 0, 0, 4, 1]}[;
[fg169,169,169;[bg;2016-08-03 18:10:33.396 [Debug] [com.github.shogo4405.lf.NetSocket.network] [NetSocket.swift:139] stream(_:handleEvent:) > eventCode: NSStreamEvent(rawValue: 2)[;
[fg211,211,211;[bg;2016-08-03 18:10:33.411 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPConnection.swift:405] listen(_:bytes:) > RTMPChunk{size:256,type:Zero,streamId:3,message:Optional(RTMPCommandMessage{type:AMF0Command,length:240,streamId:0,timestamp:0,payload:[2, 0, 7, 95, 114, 101, 115, 117, 108, 116, 0, 63, 240, 0, 0, 0, 0, 0, 0, 3, 0, 6, 102, 109, 115, 86, 101, 114, 2, 0, 13, 70, 77, 83, 47, 51, 44, 53, 44, 51, 44, 56, 50, 52, 0, 12, 99, 97, 112, 97, 98, 105, 108, 105, 116, 105, 101, 115, 0, 64, 95, 192, 0, 0, 0, 0, 0, 0, 4, 109, 111, 100, 101, 0, 63, 240, 0, 0, 0, 0, 0, 0, 0, 0, 9, 3, 0, 5, 108, 101, 118, 101, 108, 2, 0, 6, 115, 116, 97, 116, 117, 115, 0, 4, 99, 111, 100, 101, 2, 0, 29, 78, 101, 116, 67, 111, 110, 110, 101, 99, 116, 105, 111, 110, 46, 67, 111, 110, 110, 101, 99, 116, 46, 83, 117, 99, 99, 101, 115, 115, 0, 11, 100, 101, 115, 99, 114, 105, 112, 116, 105, 111, 110, 2, 0, 21, 67, 111, 110, 110, 101, 99, 116, 105, 111, 110, 32, 115, 117, 99, 99, 101, 101, 100, 101, 100, 46, 0, 14, 111, 98, 106, 101, 99, 116, 69, 110, 99, 111, 100, 105, 110, 103, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 100, 97, 116, 97, 8, 0, 0, 0, 1, 0, 7, 118, 101, 114, 115, 105, 111, 110, 2, 0, 9, 51, 44, 53, 44, 51, 44, 56, 50, 52, 0, 0, 9, 0, 0, 9],objectEncoding:0,commandName:_result,transactionId:1,commandObject:Optional(["fmsVer": Optional("FMS/3,5,3,824"), "capabilities": Optional(127.0), "mode": Optional(1.0)]),arguments:[Optional(["level": Optional("status"), "code": Optional("NetConnection.Connect.Success"), "objectEncoding": Optional(0.0), "data": Optional([Optional(undefined)]), "description": Optional("Connection succeeded.")])],serializer:AMF0Serializer{bytes:[],position:0,reference:lf.AMFReference}}),fragmented:false,_bytes:[3, 0, 0, 0, 0, 0, 240, 20, 0, 0, 0, 0]}[;
[fg211,211,211;[bg;2016-08-03 18:10:33.480 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPSocket.swift:65] doOutput(chunk:) > RTMPChunk{size:0,type:One,streamId:2,message:Optional(RTMPSetChunkSizeMessage{type:ChunkSize,length:12,streamId:0,timestamp:0,payload:[0, 0, 32, 0],size:8192}),fragmented:false,_bytes:[66, 0, 0, 0, 0, 0, 4, 1]}[;
[fg211,211,211;[bg;2016-08-03 18:10:33.536 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPSocket.swift:65] doOutput(chunk:) > RTMPChunk{size:0,type:Zero,streamId:3,message:Optional(RTMPCommandMessage{type:AMF0Command,length:37,streamId:0,timestamp:0,payload:[2, 0, 12, 99, 114, 101, 97, 116, 101, 83, 116, 114, 101, 97, 109, 0, 64, 0, 0, 0, 0, 0, 0, 0, 5],objectEncoding:0,commandName:createStream,transactionId:2,commandObject:nil,arguments:[],serializer:AMF0Serializer{bytes:[],position:0,reference:lf.AMFReference}}),fragmented:false,_bytes:[3, 0, 0, 0, 0, 0, 25, 20, 0, 0, 0, 0]}[;
[fg211,211,211;[bg;2016-08-03 18:10:33.539 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPConnection.swift:405] listen(_:bytes:) > RTMPChunk{size:256,type:One,streamId:3,message:Optional(RTMPCommandMessage{type:AMF0Command,length:21,streamId:0,timestamp:0,payload:[2, 0, 8, 111, 110, 66, 87, 68, 111, 110, 101, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5],objectEncoding:0,commandName:onBWDone,transactionId:0,commandObject:Optional([:]),arguments:[],serializer:AMF0Serializer{bytes:[],position:0,reference:lf.AMFReference}}),fragmented:false,_bytes:[67, 0, 0, 0, 0, 0, 21, 20]}[;
[fg169,169,169;[bg;2016-08-03 18:10:33.585 [Debug] [com.github.shogo4405.lf.NetSocket.network] [NetSocket.swift:139] stream(_:handleEvent:) > eventCode: NSStreamEvent(rawValue: 4)[;
[fg169,169,169;[bg;2016-08-03 18:10:42.332 [Debug] [com.github.shogo4405.lf.NetSocket.network] [NetSocket.swift:139] stream(_:handleEvent:) > eventCode: NSStreamEvent(rawValue: 2)[;
[fg211,211,211;[bg;2016-08-03 18:10:42.334 [Verbose] [com.github.shogo4405.lf.NetSocket.network] [RTMPConnection.swift:405] listen(_:bytes:) > RTMPChunk{size:256,type:One,streamId:3,message:Optional(RTMPCommandMessage{type:AMF0Command,length:29,streamId:0,timestamp:0,payload:[2, 0, 7, 95, 114, 101, 115, 117, 108, 116, 0, 64, 0, 0, 0, 0, 0, 0, 0, 5, 0, 63, 240, 0, 0, 0, 0, 0, 0],objectEncoding:0,commandName:_result,transactionId:2,commandObject:Optional([:]),arguments:[Optional(1.0)],serializer:AMF0Serializer{bytes:[],position:0,reference:lf.AMFReference}}),fragmented:false,_bytes:[67, 0, 0, 0, 0, 0, 29, 20]}[;
[fg169,169,169;[bg;2016-08-03 18:10:42.338 [Debug] [com.github.shogo4405.lf.NetSocket.network] [NetSocket.swift:139] stream(_:handleEvent:) > eventCode: NSStreamEvent(rawValue: 4)[;
[fg211,211,211;[bg;2016-08-03 18:10:42.339 [Verbose] [com.github.shogo4405.lf.Stream.lock] [RTMPSocket.swift:65] doOutput(chunk:) > RTMPChunk{size:0,type:Zero,streamId:3,message:Optional(RTMPCommandMessage{type:AMF0Command,length:41,streamId:0,timestamp:0,payload:[2, 0, 9, 70, 67, 80, 117, 98, 108, 105, 115, 104, 0, 64, 8, 0, 0, 0, 0, 0, 0, 5, 2, 0, 4, 108, 105, 118, 101],objectEncoding:0,commandName:FCPublish,transactionId:3,commandObject:nil,arguments:[Optional("live")],serializer:AMF0Serializer{bytes:[],position:0,reference:lf.AMFReference}}),fragmented:false,_bytes:[3, 0, 0, 0, 0, 0, 29, 20, 0, 0, 0, 0]}[;
[fg211,211,211;[bg;2016-08-03 18:10:42.342 [Verbose] [com.github.shogo4405.lf.Stream.lock] [RTMPSocket.swift:65] doOutput(chunk:) > RTMPChunk{size:0,type:Zero,streamId:4,message:Optional(RTMPCommandMessage{type:AMF0Command,length:46,streamId:1,timestamp:0,payload:[2, 0, 7, 112, 117, 98, 108, 105, 115, 104, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 2, 0, 4, 108, 105, 118, 101, 2, 0, 4, 108, 105, 118, 101],objectEncoding:0,commandName:publish,transactionId:0,commandObject:nil,arguments:[Optional("live"), Optional("live")],serializer:AMF0Serializer{bytes:[],position:0,reference:lf.AMFReference}}),fragmented:false,_bytes:[4, 0, 0, 0, 0, 0, 34, 20, 1, 0, 0, 0]}[;
```

**Comment by geitoodevs** [comment]
In fact, I'm using live Facebook:

rtmp://rtmp-api.facebook.com:80/rtmp/

The point is that If I use wifi network, everything is ok, but when I turn off the wifi and I try to connect,  the "preview" of the video is not shown and the last two messages from the "Wifi log" are not shown also.

By the way, How can I know if I am being blocking for instance by my provider which is Vodafone? I mean, Is there a way to get an error just to inform when the streaming is not working due to a port blocking? (That's because now I cannot notice in my app if the streaming is failing in this part or not)

Can I change the port just for testing and to be sure that is a blocking port issue?

**Comment by shogo4405** [comment]
> Can I change the port just for testing

No way. Facebook Live is fixed a port 80. iOS is auto.
> and to be sure that is a blocking port issue?

I don't think so a blocking port issue. thought you are using an original server (Wowza) 

https://www.facebook.com/facebookmedia/get-started/live
umm... stream key is at once.

I can publish to Facebook Live by 4g network (docomo/Japan). 

1.  get a stream key
1.  open preview window
1.  publish to Facebook Live (wifi) 
1.  Turn off Wifi. 
1.  close preview window.
1.  get a stream key once again
1.  open preview window.
1.  publish to Facebook Live(4g)

![img_2182](https://cloud.githubusercontent.com/assets/810189/17374744/01b05262-59e9-11e6-827a-13a1cbcdb702.png)

**Comment by geitoodevs** [comment]
I tried first to connect with wifi by generating a new stream key and it works and the same for mobile data but it's impossible -.-'. It doesn't like 4G. 

I read that they (Facebook API) have a specific list of requirements of the video to be sent.
https://developers.facebook.com/docs/videos/live-video-api

Could you tell me which is your configuration? So I can simulate the same case. Here I'm using Vodafone/Spain provider for 4G.

I tried to publish the video using the Graph API (FacebookSDK) and not the dialog as I can see in your example.  
Anyway I'll try to publish through the dialog when I have some time and I will come back to tell you if it works. The most strange thing is why It works by wifi =/

Thanks,

**Comment by shogo4405** [comment]
> Could you tell me which is your configuration? 
Sure. I tested same Exsamples/iOS configuration. I want to send you that the Facebook Live configuration. Please tell me your emai address. my email address is shogo4405 at gmail.com

**Comment by geitoodevs** [comment]
Mine is miguel at appsonline.es

**Comment by shogo4405** [comment]
@geitoodevs 
Hi. How about RTMPS? I implemented RTMPS #88.
4G Provider can't read TSL communications.
That just might work...

```swift
rtmpConnection.connect("rtmps://rtmp-api.facebook.com:443/rtmp/")
rtmpStream.publish("your stream key")
```

**Comment by geitoodevs** [comment]
It worked!!! Thank you so much!

2016-08-11 10:47 GMT+01:00 shogo4405 <notifications@github.com>:
